ubi_manifest/worker/tasks/content_audit.py

```diff
@@ -0,0 +1,165 @@
+import logging
+import os
+from concurrent.futures import Future
+from typing import Any
+
+from pubtools.pulplib import Criteria, YumRepository
+from ubiconfig import UbiConfig
+
+from ubi_manifest.worker.tasks.celery import app
+from ubi_manifest.worker.tasks.depsolver.models import UbiUnit
+from ubi_manifest.worker.tasks.depsolver.pulp_queries import (
+    search_modulemd_defaults,
+    search_modulemds,
+    search_rpms,
+)
+from ubi_manifest.worker.tasks.depsolver.ubi_config import UbiConfigLoader
+from ubi_manifest.worker.tasks.depsolver.utils import (
+    ContentConfigMissing,
+    create_or_criteria,
+    get_criteria_for_modules,
+    is_blacklisted,
+    make_pulp_client,
+    parse_blacklist_config,
+)
+
+_LOG = logging.getLogger(__name__)
+MAX_WORKERS = int(os.getenv("UBI_MANIFEST_DEPSOLVER_WORKERS", "8"))
+
+
+@app.task  # type: ignore [misc]  # ignore untyped decorator
+def content_audit_task(content_config_url: str) -> None:
+    """
+    This task checks that all available content is up-to-date, that whitelisted
+    content is present, and that blacklisted content is absent.
+    """
+
+    config_loader = UbiConfigLoader(content_config_url)
+
+    with make_pulp_client(app.conf) as client:
+        ubi_repos = client.search_repository(
+            Criteria.with_field("ubi_population", True)
+        )
+
+        for ubi_repo in ubi_repos.result():
+            ubi_config = _get_ubi_config(config_loader, ubi_repo.ubi_config_version)
+            ubi_whitelist = {pkg.name for pkg in ubi_config.packages.whitelist}
+            ubi_blacklist = parse_blacklist_config(ubi_config)
+
+            for input_repo in client.search_repository(
+                Criteria.with_id(ubi_repo.population_sources)
+            ).result():
+                seen_pkgs = set()
+
+                # check that all content is up-to-date
+                for ubi_rpm in ubi_repo.rpm_content:
```

Comment:
Also as I stated before, it's not ideal to continue and query RPMs one by one, this will lead to many requests and slow performance.

I guess you can something like:
* asynchronously query for all units in output repos
* get required criteria for the units you need query from input repos
* query for all input units
* filter them
* compare them
* log if necessary

Reply:
I considered accumulating the units and then matching them with nested loop but thought we'd need all the units from each repo anyway and that this might actually be faster. I'll revert that. However, I'm not clear on what the input unit search criteria would be. Is it expected that input repos contain more than what's in output repos to necessitate filtering?